server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -230,40 +225,41 @@ private void getFromTranslog(
         ClusterStateObserver observer,
         ActionListener<GetResponse> listener
     ) {
-        tryGetFromTranslog(request, indexShard, state, listener.delegateResponse((l, e) -> {
-            final var cause = ExceptionsHelper.unwrapCause(e);
-            logger.debug("get_from_translog failed", cause);
-            if (cause instanceof ShardNotFoundException || cause instanceof IndexNotFoundException) {
-                logger.debug("retrying get_from_translog");
-                observer.waitForNextChange(new ClusterStateObserver.Listener() {
-                    @Override
-                    public void onNewClusterState(ClusterState state) {
-                        getFromTranslog(request, indexShard, state, observer, l);
-                    }
+        try {
+            final var node = getCurrentNodeOfPrimary(state, indexShard.shardId());
+            final var retryingListener = listener.delegateResponse((l, e) -> {
```

Comment:
Nit: this can be lifted outside of the try/catch block. I find that makes it easiser to read.